### PR TITLE
Initial implementation for pickle5 support

### PR DESCRIPTION
**Pull request by suquark**
## Why are these changes needed?

This one of the plans to refactor the serialization part.

- [x] Pass all tests on python3.8
- [x] Backport `pickle5`
- [x] Check compatibility of `cloudpickle_fast.py`
- [ ] ~~Direct Arrow zero-copy support for `PickleBuffer`~~(should implement in the future)

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://ray.readthedocs.io/en/latest/.

### CODE REVIEW

**Comment by pcmoritz** [comment]
This looks great @suquark! Can we make it so there is a flag that uses one or the other codepath (the default should be the old one) so we can merge this earlier and start experimenting with it?

**Comment by pcmoritz** [comment]
@suquark There are some more linting errors, can you also fix them? Let's merge the PR as it is and address the other issues in follow up PRs :)

**Comment by suquark** [comment]
@pcmoritz Sorry for my late response. I am still working on it, it's not ready for review.

**Comment by pcmoritz** [comment]
@richardliaw has also created a PR to update cloudpickle, you should coordinate to avoid conflicts (also make sure the version gets updated), see https://github.com/ray-project/ray/pull/5643 :)

**Comment by suquark** [comment]
@pcmoritz Thanks for your mention! I am also using the lastest cloudpickle but slightly modified. I think I should wait #5643 to merge so that this PR can work as a small patch.

**Comment by pcmoritz** [comment]
@suquark Yeah, makes sense! :)

**Comment by pcmoritz** [comment]
@suquark Looks like the other PR has some problem to pass Jenkins. We shouldn't have this PR be blocked on the other one, how about we just remove the update of cloudpickle.py from this PR, so we can update it separately in a different PR?

**Comment by suquark** [comment]
@pcmoritz Yep, but then I have to drop the Python3.8 cloudpickle_fast support (it depends on the lastest cloudpickle). But I think it is totally fine to do so since we will not support python3.8 recently.

**Comment by suquark** [comment]
The jenkins build failed with:

```python
    Autodetected HDF5 1.10.2
    ********************************************************************************
                           Summary of the h5py configuration
    
        Path to HDF5: None
        HDF5 Version: '1.10.2'
         MPI Enabled: False
    Rebuild Required: True
    
    ********************************************************************************
    Executing cythonize()
    h5py requires pkg-config unless the HDF5 path is explicitly specified
    error: pkg-config probably not installed: FileNotFoundError(2, "No such file or directory: 'pkg-config'")
```

**Comment by suquark** [comment]
It seems that all tests have failed due to h5py. Not sure what has happened.

**Comment by suquark** [comment]
jenkins retest it please

**Comment by suquark** [comment]
close and reopen the PR to re-trigger all tests

**Comment by pcmoritz** [comment]
Now that we have https://github.com/ray-project/ray/pull/5643 merged, can you rebase this PR and we can get it merged as well?

**Comment by suquark** [comment]
@pcmoritz it's ready to review and merge

File: python/ray/cloudpickle/cloudpickle.py
Comment (pcmoritz) [inline python/ray/cloudpickle/cloudpickle.py]: This should only be needed for cloudpickle_fast, right? Let's focus on supporting that codepath, so it doesn't break other things.

**Comment by pcmoritz** [comment]
Thanks, we need to make sure that cloudpickle as well as cloudpickle_fast are unchanged from upstream. We can't start to diverge from them, otherwise it will be very hard to maintain our fork in the future (there are frequent changes in cloudpickle).

If they need to be changed, we need to get the changes upstream. If possible let's try to avoid that (there is a way to not change them by re-running the CloudPickler.__init__ as in the code I shared with you, right?)

**Comment by suquark** [comment]
@pcmoritz Yep, but that code doesn't work somehow.

**Comment by pcmoritz** [comment]
@suquark Do you know what is going wrong?

File: python/ray/cloudpickle/cloudpickle.py
Comment (suquark) [reply python/ray/cloudpickle/cloudpickle.py]: If we we have pickle5 enabled (python3 && python<3.8 && pickle5 installed), then we will try to reuse the old cloudpickle

**Comment by pcmoritz** [comment]
It's fine if we cannot support the old cloudpickle with pickle5 for now I think (the old cloudpickle can be quite slow like 20x slower already on python objects since it uses the python pickle implementation)

**Comment by suquark** [comment]
@pcmoritz I am not sure. But it could because of how python3.8 deals with the C++ hook function in cloudpickle_fast. I haven't tried on `pickle5-backport`, but it is likely to work because `pickle5-backport` haven't implemented the hook function. BTW, the hook function is critical for performance. It is not possible to make use of cPickle without the hook function (everything falls back to python-based pickle). I also created this issue for pickle5-backport: https://github.com/pitrou/pickle5-backport/issues/9

**Comment by suquark** [comment]
@pcmoritz  yes, but I can create a PR to make the old cloudpickle 20x faster with `pickle5-backport`

**Comment by suquark** [comment]
Here's the cloudpickle PR: https://github.com/cloudpipe/cloudpickle/pull/308

**Comment by pcmoritz** [comment]
Thanks for doing it, looks great!

**Review by pcmoritz — APPROVED** [review]
Can be merged when tests pass

**Pull request merged**
